docs/architecture.md

```diff
@@ -18,3 +18,94 @@
 
 ## Back-end architecture
+
+![Diagram](images/architecture-back-end.drawio.svg)
+
+### Dataset files
+
+All persistent data is stored on the back-end server as JSON files, in the following folder structure
+as seen from the SERVER_DATA_ROOT location:
+
+```
+├── datasets
+│   ├── some-dataset
+│   │   ├── locations.json (array of lng-lat coordinates for each initiative)
+│   │   ├── searchable.json (array of the property values and searchable strings for each initiative)
+│   │   ├── initiatives
+│   │   |   ├── 0.json (full info of first initiative in the above aggregate JSONs)
+│   │   |   ├── 1.json
+│   │   |   ├── ...
+│   ├── other-dataset
+│   │   ├── ...
+│   ├── ...
+```
+
+Additionally, for each map variant, there's a `config.json`, which contains config for displaying the map in the UI, including the vocabs (translations of data IDs), default sidebar panel, and popup appearance.
+
+### Example file contents
+
+`locations.json`:
+
+```
+[ [1.21419, 50.45254], [0.21002, 49.33954], … ]
+```
+
+`searchable.json`:
+
+```
+[ { "coun": "coun:GB", "sz": "sz:small", "searchString": "some co-op 2 green lane london n4 9qr" }, … ]
+```
+
+`0.json`:
+
+```
+{ name: "Some Co-op", "desc": "A co-op that sells stuff", "lng": 1.21419, "lat": 50.45254, "coun": "coun:GB", "sz": "sz:small", ... }
+```
+
+`config.json`:
+
+```
+{
+  "prefixes": {
+    "https://example.com/sizes/1.1/": "sz",
+    ...
+  },
+  "vocabs": {
+    "sz:": {
+      "EN": {
+        "title": "Sizes",
+        "terms": {
+          "sz:large": "Large",
+          "sz:medium": "Medium",
+          "sz:small": "Small"
+        }
+      }
+    },
+    ...
+  },
+  "popups": { "field": "sz", "style": "text" }, { "field": "websites", "style": "clickable-list" },
+  "ui": { ... },
+  ...
+}
+
+```
+
+### Data generation
+
+These directories of JSONs, including the searchable strings in the `searchable.json` files, need to be pre-generated by a script. This script will be written in JS/TS and live in the monorepo, to be run on the back-end server.
+
+The script will take the full data CSV for a map (generated by the data factory) as inputs, and write the full data into the required JSON files in the directory structure specified above.
+
+#### Note:
+
+We will need to manually copy the `standard.csv` from the data factory server to the back-end. Maybe in the future, the data factory pipeline can be enhanced to write the JSON files to the back-end server so that no manual duplication is necessary (and maybe we can eventually get rid of the separate data server altogether).
+
+### Dataset instances
+
+- For each dataset available in the `datasets` directory on server start, a dataset instance is created
+  by the Dataset service. Each Dataset instance has a:
+  - `searchable` property, which is just the `searchable.json` loaded as an in-memory object
+  - `getItem` method
+  - `getConfig` method, which includes the vocabs
```

Comment:
This could be be streamed by the back-end too, in which case it would need a path as below, or a method which returns a stream.

Reply:
I'm not sure if streaming works with runtime Zod response validation, I'm guessing not. So I thought it made sense to not stream in this case, unless we expect the config files to get very large. Could always change it in the future if needed

Reply:
Yes, streaming bypasses ts-rest's on-the-fly Zod validation, I think. However, this is not to say that such streamed files couldn't be validated on creation, and I think they should be. Do we benefit by validating every request after that? Not sure, suspect not much.

But overall this is just a suggestion, as you say the configs will be relatively small.